File: Example/javaapp/src/main/java/com/stripe/example/javaapp/fragment/offline/OfflinePaymentsLogFragment.java
package com.stripe.example.javaapp.fragment.offline;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.stripe.example.javaapp.MainActivity;
import com.stripe.example.javaapp.NavigationListener;
import com.stripe.example.javaapp.R;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class OfflinePaymentsLogFragment extends Fragment {
    @NotNull
    public static final String TAG = "com.stripe.example.fragment.OfflinePaymentsLogFragment";
    private OfflinePaymentsLogAdapter adapter;
    @Override
    @Nullable
    public View onCreateView(
            @NotNull LayoutInflater inflater,
            @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState
    ) {
        return inflater.inflate(R.layout.fragment_offline_payments_log, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @androidx.annotation.Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        view.findViewById(R.id.done_button).setOnClickListener(v -> {
            FragmentActivity activity = getActivity();
            if (activity instanceof NavigationListener) {
                ((MainActivity) activity).onRequestExitWorkflow();
            }
        });

        final RecyclerView logRecyclerView = view.findViewById(R.id.offline_logs_recycler_view);
        logRecyclerView.setLayoutManager(new LinearLayoutManager(requireContext()));
        adapter = new OfflinePaymentsLogAdapter();
        logRecyclerView.setAdapter(adapter);

        FragmentActivity activity = requireActivity();
        if (activity instanceof MainActivity) {
            ((MainActivity) activity).offlineModeHandler.liveLogs.observe(getViewLifecycleOwner(),
                logs -> adapter.updateOfflineLogs(logs)
            );
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();

        FragmentActivity activity = requireActivity();
        if (activity instanceof MainActivity) {
            ((MainActivity) activity).offlineModeHandler.liveLogs.removeObservers(getViewLifecycleOwner());
        }
    }
}

File: Example/kotlinapp/src/main/java/com/stripe/example/fragment/offline/OfflinePaymentsLogFragment.kt
package com.stripe.example.fragment.offline

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.stripe.example.NavigationListener
import com.stripe.example.R
import com.stripe.example.TerminalOfflineListener
import com.stripe.example.fragment.launchAndRepeatWithViewLifecycle
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch

class OfflinePaymentsLogFragment : Fragment() {
    private val adapter: OfflinePaymentsLogAdapter = OfflinePaymentsLogAdapter()
    override fun onCreateView(
            inflater: LayoutInflater,
            container: ViewGroup?,
            savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_offline_payments_log, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        launchAndRepeatWithViewLifecycle(Lifecycle.State.RESUMED) {
            launch {
                TerminalOfflineListener.offlineEventsFlow
                        .collectLatest {
                            adapter.updateOfflineLogs(it)
                        }
            }
        }

        view.findViewById<View>(R.id.done_button).setOnClickListener {
            (activity as? NavigationListener)?.onRequestExitWorkflow()
        }

        view.findViewById<RecyclerView>(R.id.offline_logs_recycler_view)?.run {
            this.layoutManager = LinearLayoutManager(requireContext())
            this.adapter = this@OfflinePaymentsLogFragment.adapter
        }
    }

    companion object {
        const val TAG = "com.stripe.example.fragment.OfflinePaymentsLogFragment"
    }
}
